src/k8s/pkg/client/dqlite/remove.go

```diff
@@ -27,21 +31,49 @@ func (c *Client) RemoveNodeByAddress(ctx context.Context, address string) error
 
 		case member.Address != address && member.Role == Voter:
 			clusterHasOtherVoters = true
+
+		case member.Address != address && member.Role == Spare:
+			// Mark as fail-over node. This is only used in a two node setup,
+			// where the leader node is removed. The fail-over node will be
+			// promoted to leader before the existing leader is removed.
+			failOverMemberID = &member.ID
 		}
 	}
 
 	if !memberExists {
 		return fmt.Errorf("cluster does not have a node with address %v", address)
 	}
 
-	// TODO: consider using client.Transfer() for a different node to become leader
 	if !clusterHasOtherVoters {
-		return fmt.Errorf("not removing node because there are no other voter members")
-	}
+		if failOverMemberID == nil {
+			// This normally should not happen. There should always be a backup node, except
+			// if one tries to remove the last node in the cluster.
+			return fmt.Errorf("cannot transfer dqlite leadership as there is no remaining spare node")
+		}
 
-	if err := client.Remove(ctx, memberToRemove.ID); err != nil {
-		return fmt.Errorf("failed to remove node %#v from dqlite cluster: %w", memberToRemove, err)
+		// Leadership can only be transfered to a voter or standby node.
+		// Therefore the remaining node in the cluster needs to be promoted first.
+		if err := client.Assign(ctx, *failOverMemberID, Voter); err != nil {
+			return fmt.Errorf("failed to assign voter role to %d: %w", *failOverMemberID, err)
+		}
+		// Transfer leadership to remaining node in cluster.
+		if err := client.Transfer(ctx, *failOverMemberID); err != nil {
+			return fmt.Errorf("failed to transfer leadership to %d: %w", *failOverMemberID, err)
+		}
+		// Recreate client to point to the new leader.
+		client, err = c.clientGetter(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to create dqlite client: %w", err)
+		}
 	}
 
-	return nil
+	// Remove the node from the cluster. Retry as the leadership transfer might still be in progress.
+	// For a large database this might take some time.
+	return control.RetryFor(10, func() error {
+		if err := client.Remove(ctx, memberToRemove.ID); err != nil {
```

Comment:
does this `client.Remove` fail after a specific amount of time?  I'm trying to count the seconds this loop would run

* 10 retries
* undetermined number of seconds for the `client.Remove`
* 5 seconds of sleep back off

could be as QUICK as 0 seconds on success, to heat death of the universe if `client.Remove` never times out

was the intent instead to retry 10 times with 5 seconds between calls with some maximum timeout of 60s?

Reply:
Good point, the timeout of `client.Remove` is not something we have control over. 
However, we provide a context that, if cancelled, will cause the `client.Remove` to stop.

However, the `RetryFor` function did not check if the context was already cancelled before retrying which causes the function the hang as it retries again.

I added a fix for that and a unittests to verify the behavior.